core/src/blocks/curl.rs

```diff
@@ -0,0 +1,239 @@
+use url::{Host, Url};
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Error {
+    pub error: String,
+}
+
+#[derive(Clone)]
+pub struct Curl {
+    method: String,
+    url: String,
+    headers_code: String,
+    body_code: String,
+}
+
+impl Curl {
+    pub fn parse(block_pair: Pair<Rule>) -> Result<Self> {
+        let mut method: Option<String> = None;
+        let mut url: Option<String> = None;
+        let mut headers_code: Option<String> = None;
+        let mut body_code: Option<String> = None;
+
+        for pair in block_pair.into_inner() {
+            match pair.as_rule() {
+                Rule::pair => {
+                    let (key, value) = parse_pair(pair)?;
+                    match key.as_str() {
+                        "method" => method = Some(value),
+                        "url" => url = Some(value),
+                        "headers_code" => headers_code = Some(value),
+                        "body_code" => body_code = Some(value),
+                        _ => Err(anyhow!("Unexpected `{}` in `curl` block", key))?,
+                    }
+                }
+                Rule::expected => Err(anyhow!("`expected` is not yet supported in `curl` block"))?,
+                _ => unreachable!(),
+            }
+        }
+
+        if !method.is_some() {
+            Err(anyhow!("Missing required `method` in `curl` block"))?;
+        }
+        if !url.is_some() {
+            Err(anyhow!("Missing required `url` in `curl` block"))?;
+        }
+        if !headers_code.is_some() {
+            Err(anyhow!("Missing required `headers_code` in `curl` block"))?;
+        }
+        if !body_code.is_some() {
+            Err(anyhow!("Missing required `body_code` in `curl` block"))?;
+        }
+
+        Ok(Curl {
+            method: method.unwrap(),
+            url: url.unwrap(),
+            headers_code: headers_code.unwrap(),
+            body_code: body_code.unwrap(),
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct CurlResult {
+    status: u16,
+    body: Option<serde_json::Value>,
+    error: Option<String>,
+}
+
+#[async_trait]
+impl Block for Curl {
+    fn block_type(&self) -> BlockType {
+        BlockType::Curl
+    }
+
+    fn inner_hash(&self) -> String {
+        let mut hasher = blake3::Hasher::new();
+        hasher.update("curl".as_bytes());
+        hasher.update(self.method.as_bytes());
+        hasher.update(self.url.as_bytes());
+        hasher.update(self.headers_code.as_bytes());
+        hasher.update(self.body_code.as_bytes());
+        format!("{}", hasher.finalize().to_hex())
+    }
+
+    async fn execute(&self, _name: &str, env: &Env) -> Result<Value> {
+        let method = match self.method.as_str() {
+            "GET" => Method::GET,
+            "POST" => Method::POST,
+            "PUT" => Method::PUT,
+            "PATCH" => Method::PATCH,
+            _ => Err(anyhow!(
+                "Invalid method {}, supported methods are GET, POST, PUT.",
```

Comment:
This is a tiny nit, but it might be better to do this checking earlier on during the parse and then store the variable as a Method.